**Python API Tutorial: Getting Started with APIs** <br>
Link: https://www.dataquest.io/blog/python-api-tutorial/

In [2]:
import requests

In [4]:
response = requests.get("http://api.open-notify.org/this-api-doesnt-exist")

In [6]:
print(response.status_code)

404


API Status Codes
Status codes are returned with every request that is made to a web server. Status codes indicate information about what happened with a request. Here are some codes that are relevant to GET requests:

200: Everything went okay, and the result has been returned (if any). <br>
301: The server is redirecting you to a different endpoint. This can happen when a company switches domain names, or an endpoint name is changed. <br>
400: The server thinks you made a bad request. This can happen when you don’t send along the right data, among other things. <br>
401: The server thinks you’re not authenticated. Many APIs require login ccredentials, so this happens when you don’t send the right credentials to access an API. <br>
403: The resource you’re trying to access is forbidden: you don’t have the right permissions to see it. <br>
404: The resource you tried to access wasn’t found on the server. <br>
503: The server is not ready to handle the request. <br>

You might notice that all of the status codes that begin with a ‘4’ indicate some sort of error. The first number of status codes indicate their categorization. This is useful — you can know that if your status code starts with a ‘2’ it was successful and if it starts with a ‘4’ or ‘5’ there was an error. If you’re interested you can read more about status codes



In [7]:
response = requests.get("http://api.open-notify.org/astros.json")
print(response.status_code)

200


In [8]:
print(response.json())

{'people': [{'name': 'Christina Koch', 'craft': 'ISS'}, {'name': 'Alexander Skvortsov', 'craft': 'ISS'}, {'name': 'Luca Parmitano', 'craft': 'ISS'}, {'name': 'Andrew Morgan', 'craft': 'ISS'}, {'name': 'Oleg Skripochka', 'craft': 'ISS'}, {'name': 'Jessica Meir', 'craft': 'ISS'}], 'number': 6, 'message': 'success'}


In [9]:
import json

def jprint(obj):
    # create a formatted string of the Python JSON object
    text = json.dumps(obj, sort_keys=True, indent=4)
    print(text)

jprint(response.json())

{
    "message": "success",
    "number": 6,
    "people": [
        {
            "craft": "ISS",
            "name": "Christina Koch"
        },
        {
            "craft": "ISS",
            "name": "Alexander Skvortsov"
        },
        {
            "craft": "ISS",
            "name": "Luca Parmitano"
        },
        {
            "craft": "ISS",
            "name": "Andrew Morgan"
        },
        {
            "craft": "ISS",
            "name": "Oleg Skripochka"
        },
        {
            "craft": "ISS",
            "name": "Jessica Meir"
        }
    ]
}


In [10]:
parameters = {
    "lat": 40.71,
    "lon": -74
}

In [11]:
response = requests.get("http://api.open-notify.org/iss-pass.json", params=parameters)

jprint(response.json())

{
    "message": "success",
    "request": {
        "altitude": 100,
        "datetime": 1571414652,
        "latitude": 40.71,
        "longitude": -74.0,
        "passes": 5
    },
    "response": [
        {
            "duration": 583,
            "risetime": 1571419717
        },
        {
            "duration": 647,
            "risetime": 1571425521
        },
        {
            "duration": 565,
            "risetime": 1571431347
        },
        {
            "duration": 603,
            "risetime": 1571485674
        },
        {
            "duration": 645,
            "risetime": 1571491446
        }
    ]
}


**Let’s extract the pass times from our JSON object:**

In [12]:
pass_times = response.json()['response']
jprint(pass_times)

[
    {
        "duration": 583,
        "risetime": 1571419717
    },
    {
        "duration": 647,
        "risetime": 1571425521
    },
    {
        "duration": 565,
        "risetime": 1571431347
    },
    {
        "duration": 603,
        "risetime": 1571485674
    },
    {
        "duration": 645,
        "risetime": 1571491446
    }
]


**Next we’ll usea loop to extract just the five risetime values:**

In [13]:
risetimes = []

for d in pass_times:
    time = d['risetime']
    risetimes.append(time)

print(risetimes)

[1571419717, 1571425521, 1571431347, 1571485674, 1571491446]


**These times are difficult to understand – they are in a format known as timestamp or epoch. Essentially the time is measured in the number of seconds since January 1st 1970. We can use the Python datetime.fromtimestamp() method to convert these into easier to understand times:**

In [14]:
from datetime import datetime

times = []

for rt in risetimes:
    time = datetime.fromtimestamp(rt)
    times.append(time)
    print(time)

2019-10-18 13:28:37
2019-10-18 15:05:21
2019-10-18 16:42:27
2019-10-19 07:47:54
2019-10-19 09:24:06


In this tutorial, we learned:

What an API is
Types of requests and response codes
How to make a get request
How to make a request with parameters
How to display and extract JSON data from an API

**Remember that key to each time we used the API was to carefully read the API documentation and use that to understand what request to make and what parameters to provide.**